Fix go back history on forms

paulfalgout opened the pull request:
Shortcut Story ID: [sc-34125]

So this latest history service is working far far better than the previous one, but 🤦 I wasn't removing the current route before going back in this release, so it was going to the current route.

In any case I snuck in an empty history handler which defaults forms to go to the patient's dashboard if there was no previous history recorded @shadowhand   seems to work.

paulfalgout committed:
Fix history goback

Newest iteration was keeping great track of history, but it wasn’t removing the current route so it would “go back” to itself.

I also added an `handleEmptyHistory` optional handler for if there’s no previous history.  This fixes @woody’s concern with direct links to forms not going back to anywhere useful.

shadowhand commented on src/js/services/history.js:
```diff
@@ -48,7 +48,16 @@ export default App.extend({
+      return;
+    }
+
+    // Remove the current route from the stack
+    this._prevHistory.shift();
```
We're sure this is a good idea? Should it happen before the top `if` condition?

paulfalgout replied:
yeah so it's a _little_ awkward.. if you're directly linking to a form the `_prevHistory` will be just that form url.  so then the go back isn't go back in history.. it's just linking to the patient..  

so in theory if there was a "go back" button on the patient page, if we remove it here, that button wouldn't go back to the form, it would go to the default app route.  I think you'd want it to bounce back and forth between the two routes. 

The `onEmptyHistory` would be clearing the history instead of doing some sort of default redirect, and I don't know if that's the ideal behavior.

shadowhand commented on src/js/views/forms/form/form_views.js:
```diff
@@ -34,9 +34,14 @@ const ContextTrailView = View.extend({
   },
   onClickBack() {
-    Radio.request('history', 'go:back');
+    Radio.request('history', 'go:back', () => {
+      this.routeToPatient();
+    });
   },
   onClickDashboard() {
+    this.routeToPatient();
+  },
+  routeToPatient() {
```
```suggestion
    Radio.request('history', 'go:back', () => {
      this.onClickDashboard();
    });
  },
  onClickDashboard() {
```

paulfalgout replied:
yeah.. Ironically despite my naming of the `onEmptyHistory` on this PR an `onFoo` on a class is an indication that it's triggered by an event and we attempt not to call them directly as it's clearer when they need to be cleaned up. _but_ now I'm convinced I should rename `onEmptyHistory` :-D

shadowhand commented on src/js/services/history.js:
```diff
@@ -48,7 +48,16 @@ export default App.extend({
+    if (onEmptyHistory && this._prevHistory.length < 2) {
+      onEmptyHistory();
+      return;
+    }
```
Clever solution. 👍🏼

paulfalgout committed:
Utilize history handleEmptyHistory for default route

This change looks bigger than it is.  The additional nested function was making the nested success handler in the `listenTo` rather deep, so I used `bindEvents` to make the service handlers more similar to `stateEvents`

This change looks bigger than it is.  The additional nested function was making the nested success handler in the `listenTo` rather deep, so I used `bindEvents` to make the service handlers more similar to `stateEvents`

This change looks bigger than it is.  The additional nested function was making the nested success handler in the `listenTo` rather deep, so I used `bindEvents` to make the service handlers more similar to `stateEvents`

nmajor25 commented on src/js/views/forms/form/form_views.js:
```diff
@@ -34,9 +34,14 @@ const ContextTrailView = View.extend({
-    Radio.request('history', 'go:back');
+    Radio.request('history', 'go:back', () => {
+      this.routeToPatient();
+    });
```
It's not letting me comment on that line, but does the go back in the `PreviewView` in this file need a `handleEmptyHistory` function as well?

This line here: https://github.com/RoundingWell/care-ops-frontend/pull/944/commits/01d2176d7336ef6d2ce2b1e0772d6f6459ae4b6e#diff-e2f36acf877614e07a231a6f78b71f617ab8c1cc839269088dad7a7469d74a0bR229

paulfalgout replied:
Not sure where we'd go to necessarily.  I guess we could potentially go back to the program, but TBH the form preview is pretty broken.. it can't load context in a way that most forms are probably pretty broken to preview.

nmajor25 replied:
Okay, only thing I noticed. Things are working nicely locally.
